File: MariaDB/docker/rsg/yacc/parse.go
// Copyright 2011 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in licenses/BSD-golang.txt.

// Portions of this file are additionally subject to the following license
// and copyright.
//
// Copyright 2016 The Cockroach Authors.
//
// Use of this software is governed by the Business Source License
// included in the file licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0, included in the file
// licenses/APL.txt.

// Copied from Go's text/template/parse package and modified for yacc.

// Package yacc parses .y files.
package yacc

import (
	"fmt"
	"runtime"
)

// Tree is the representation of a single parsed file.
type Tree struct {
	Name        string // name of the template represented by the tree.
	Productions []*ProductionNode
	text        string // text parsed to create the template (or its parent)
	// Parsing only; cleared after parse.
	lex       *lexer
	token     [2]item // two-token lookahead for parser.
	peekCount int
}

// Parse parses the yacc file text with optional name.
func Parse(name, text string, dbmsName string) (t *Tree, err error) {
	t = New(name)
	t.text = text
	err = t.Parse(text, dbmsName)
	return
}

// next returns the next token.
func (t *Tree) next() item {
	//fmt.Printf("Inside next() \n\n\n")
	if t.peekCount > 0 {
		t.peekCount--
	} else {
		t.token[0] = t.lex.nextItem()
	}
	//fmt.Printf("Getting %s \n\n\n", t.token[t.peekCount].String())
	return t.token[t.peekCount]
}

// backup backs the input stream up one token.
func (t *Tree) backup() {
	t.peekCount++
}

// peek returns but does not consume the next token.
func (t *Tree) peek() item {
	if t.peekCount > 0 {
		return t.token[t.peekCount-1]
	}
	t.peekCount = 1
	t.token[0] = t.lex.nextItem()
	return t.token[0]
}

// Parsing.

// New allocates a new parse tree with the given name.
func New(name string) *Tree {
	return &Tree{
		Name: name,
	}
}

// errorf formats the error and terminates processing.
func (t *Tree) errorf(format string, args ...interface{}) {
	err := fmt.Errorf(format, args...)
	errStr := err.Error() + ", error parse: %s:%d"
	errStr = fmt.Sprintf(errStr, t.Name, t.lex.lineNumber())
	panic(errStr)
}

// expect consumes the next token and guarantees it has the required type.
func (t *Tree) expect(expected itemType, context string) item {
	token := t.next()
	if token.typ != expected {
		t.unexpected(token, context)
	}
	return token
}

// unexpected complains about the token and terminates processing.
func (t *Tree) unexpected(token item, context string) {
	t.errorf("unexpected %s in %s", token, context)
}

// recover is the handler that turns panics into returns from the top level of Parse.
func (t *Tree) recover(errp *error) {
	if e := recover(); e != nil {
		if _, ok := e.(runtime.Error); ok {
			panic(e)
		}
		if t != nil {
			t.stopParse()
		}
		*errp = e.(error)
	}
}

// startParse initializes the parser, using the lexer.
func (t *Tree) startParse(lex *lexer) {
	t.lex = lex
}

// stopParse terminates parsing.
func (t *Tree) stopParse() {
	t.lex = nil
}

// Parse parses the yacc string to construct a representation of
// the file for analysis.
func (t *Tree) Parse(text string, dbmsName string) (err error) {
	defer t.recover(&err)
	t.startParse(lex(dbmsName, text))
	t.text = text
	if dbmsName == "sqlite" {
		t.parseLemon()
	} else {
		t.parse()
	}
	t.stopParse()
	return nil
}

// parse is the top-level parser for a file.
// It runs to EOF.
func (t *Tree) parse() {
	for {
		//fmt.Printf("Inside the parse for loop. \n")
		switch token := t.next(); token.typ {
		case itemIdent:
			p := newProduction(token.pos, token.val)
			t.parseProduction(p)
			t.Productions = append(t.Productions, p)
			//fmt.Printf("From keyword: %s\n", token.val)
			//for _, expr := range p.Expressions {
			//	fmt.Printf("Getting token: %v\n", expr.Items)
			//}
			//fmt.Printf("\n\n\n")
		case itemEOF:
			//fmt.Printf("Getting itemEOF: %s\n\n\n", token.val)
			return
		}
	}
}

// parse is the top-level parser for a file.
// It runs to EOF.
func (t *Tree) parseLemon() {
	isComment := false
	for {
		switch token := t.next(); token.typ {
		case itemNL:
			isComment = false
		case itemComment:
			isComment = true
		case itemIdent:
			if isComment {
				continue
			}
			p := newProduction(token.pos, token.val)
			t.parseProductionLemon(p)
			t.Productions = append(t.Productions, p)
		case itemEOF:
			return
		}
	}
}

func (t *Tree) parseProduction(p *ProductionNode) {
	const context = "production"
	t.expect(itemColon, context)
	if t.peek().typ == itemNL {
		t.next()
	}
	expectExpr := true
	for {
		token := t.next()
		switch token.typ {
		case itemComment, itemNL:
			// ignore
		case itemPipe:
			if expectExpr {
				//t.unexpected(token, context)
				e := newExpression(token.pos)
				p.Expressions = append(p.Expressions, e)
			}
			expectExpr = true
		default:
			t.backup()
			if !expectExpr {
				return
			}
			e := newExpression(token.pos)
			t.parseExpression(e)
			p.Expressions = append(p.Expressions, e)
			expectExpr = false
		}
	}
}

func (t *Tree) parseExpression(e *ExpressionNode) {
	const context = "expression"
	for {
		switch token := t.next(); token.typ {
		case itemPipe:
			t.backup()
			return
		case itemSemicolon:
			return
		case itemNL:
			peek := t.peek().typ
			if peek == itemPipe || peek == itemNL || peek == itemSemicolon {
				return
			}
		case itemIdent:
			e.Items = append(e.Items, Item{token.val, TypToken})
		case itemLiteral:
			e.Items = append(e.Items, Item{token.val, TypLiteral})
		case itemExpr:
			e.Command = token.val
			if t.peek().typ == itemNL {
				t.next()
			}
		case itemPct, itemComment:
			// ignore
		default:
			t.unexpected(token, context)
		}
	}
}

func (t *Tree) parseProductionLemon(p *ProductionNode) {
	const context = "production"
	t.expect(itemAssign, context)
	if t.peek().typ == itemNL {
		t.next()
	}
	expectExpr := true
	for {
		token := t.next()
		switch token.typ {
		case itemComment, itemNL:
			// For the lemon rules, every grammar rule is in one line.
			// No comments are in the way.
			return
		case itemPipe:
			if expectExpr {
				t.unexpected(token, context)
			}
			expectExpr = true
		default:
			t.backup()
			if !expectExpr {
				return
			}
			e := newExpression(token.pos)
			t.parseExpressionLemon(e)
			p.Expressions = append(p.Expressions, e)
			expectExpr = false
		}
	}
}

func (t *Tree) parseExpressionLemon(e *ExpressionNode) {
	const context = "expression"
	for {
		switch token := t.next(); token.typ {
		case itemNL:
			// All expression are in one line.
			return
		case itemIdent:
			e.Items = append(e.Items, Item{token.val, TypToken})
		case itemLiteral:
			e.Items = append(e.Items, Item{token.val, TypLiteral})
		case itemExpr:
			e.Command = token.val
			if t.peek().typ == itemNL {
				t.next()
			}
			return
		case itemComment:
		// ignore
		case itemTerm:
			// If encounter the termination period, ignore all other text until end of line.
			for nextToken := t.next(); nextToken.typ != itemNL && nextToken.typ != itemEOF; {
				nextToken = t.next()
			}
			// Backup the last New Line or EOF token.
			t.backup()
			return
		case itemPipe:
			// Encounter pipe. Back it up to the parseProductionLemon.
			t.backup()
			return
		default:
			t.unexpected(token, context)
		}
	}
}

File: MySQL/docker/rsg_cpp/yacc/parse.cpp
//
// Created by XXX on 2/29/24.
//

#include "../headers/parse.h"
#include <iostream>

using std::cerr;

void Tree::start_parse(const string& text_in)
{
    this->lex = new Lexer(this->dbms_name, text_in);
    this->text = text_in;
}
void Tree::stop_parse()
{
    if (this->lex) {
        delete (this->lex);
        this->lex = nullptr;
    }
    this->text.clear();
}

int Tree::parse(const string& text_in)
{
    this->start_parse(text_in);
    this->text = text_in;
    if (this->dbms_name == "sqlite") {
        this->parse_lemon_internal();
    } else {
        this->parse_internal();
    }
    this->stop_parse();
    return 0;
}

LexTokenItem Tree::next()
{
    if (this->peek_count > 0) {
        this->peek_count--;
    } else {
        this->token[0] = this->lex->nextItem();
    }
    return this->token[this->peek_count];
}

void Tree::backup()
{
    this->peek_count++;
}

LexTokenItem Tree::peek()
{
    if (this->peek_count > 0) {
        return this->token[this->peek_count - 1];
    }
    this->peek_count = 1;
    this->token[0] = this->lex->nextItem();
    return this->token[0];
}

void Tree::parse_error(const std::string& error_str)
{
    // Encounter error, directly abort.
    cerr << "Getting parsing error: " << error_str << "\nFrom line: " << this->lex->get_cur_line_number() << "\nAbort. \n";
    abort();
}

LexTokenItem Tree::expect(LexTokenItemTyp itemTyp, const string& context)
{
    LexTokenItem cur_token = this->next();
    if (cur_token.typ != itemTyp) {
        this->unexpected(cur_token, context);
    }
    return cur_token;
}

void Tree::unexpected(LexTokenItem cur_token, const std::string& context)
{
    const string err_msg = "Unexpected token: " + cur_token.to_string() + " in " + context;
    // This function will terminate the program execution.
    this->parse_error(err_msg);
}

void Tree::parse_internal()
{
    for (;;) {
        LexTokenItem cur_token = this->next();
        if (cur_token.typ == itemIdent) {
            ProductionNode* p = get_new_production_node(cur_token.pos, cur_token.val);
            this->parse_production(p);
            this->productions.push_back(p);
        } else if (cur_token.typ == itemEOF) {
            return;
        }
    }
}

void Tree::parse_lemon_internal()
{
    bool is_comment = false;
    for (;;) {
        LexTokenItem cur_token = this->next();
        switch (cur_token.typ) {
        case itemNL: {
            is_comment = false;
            // break the switch
            break;
        }
        case itemComment: {
            is_comment = true;
            // break the switch
            break;
        }
        case itemIdent: {
            if (is_comment) {
                // break the switch
                break;
            }
            ProductionNode* p = nullptr;
            if (this->m_prods.count(cur_token.val)) {
                p = this->m_prods[cur_token.val];
            } else {
                p = get_new_production_node(cur_token.pos, cur_token.val);
            }
            this->parse_production_lemon(p);
            if (!(this->m_prods.count(cur_token.val))) {
                this->productions.push_back(p);
                this->m_prods[cur_token.val] = p;
            }
            break;
        }
        case itemEOF: {
            return;
        }
        default: {
            // Do nothing. Ignored.
        }
        }
    }
}

void Tree::parse_production(ProductionNode* p)
{
    const string context = "production";
    this->expect(itemColon, context);
    if (this->peek().typ == itemNL) {
        this->next();
    }
    bool expect_expr = true;

    for (;;) {
        const LexTokenItem cur_token = this->next();
        switch (cur_token.typ) {
        case (itemComment):
            [[fallthrough]];
        case (itemNL):
            // Ignore the current token. Continue to read the next.
            break;
        case (itemPipe): {
            if (expect_expr) {
                ExpressionNode* e = get_new_expression(cur_token.pos);
                p->all_exprs.push_back(e);
            }
            expect_expr = true;
            break;
        }
        default: {
            this->backup();
            if (!expect_expr) {
                return;
            }
            ExpressionNode* e = get_new_expression(cur_token.pos);
            this->parse_expression(e);
            p->all_exprs.push_back(e);
            expect_expr = false;
        }
        }
    }
}

void Tree::parse_production_lemon(ProductionNode* p)
{
    const string context = "production";
    this->expect(itemAssign, context);
    if (this->peek().typ == itemNL) {
        this->next();
    }
    bool expect_expr = true;

    for (;;) {
        const LexTokenItem cur_token = this->next();
        switch (cur_token.typ) {
        case (itemComment):
            [[fallthrough]];
        case (itemNL):
            // Ignore the current token. Continue to read the next.
            // For the lemon rules, every grammar rule is in one line.
            // No comments are in the way.
            return;
        case (itemPipe): {
            // TODO: This rule doesn't seem to be handled correctly.
            // FIXME: SHOULD FIX LATER.
            if (expect_expr) {
                this->unexpected(cur_token, context);
            }
            expect_expr = true;
            break;
        }
        default: {
            this->backup();
            if (!expect_expr) {
                return;
            }
            vector<ExpressionNode*> v_e = this->parse_expression_lemon(cur_token.pos);
            for (auto* e: v_e) {
                p->all_exprs.push_back(e);
            }
            expect_expr = false;
        }
        }
    }
}

void Tree::parse_expression(ExpressionNode* e)
{
    const string context = "expression";
    for (;;) {
        const LexTokenItem cur_token = this->next();
        switch (cur_token.typ) {
        case (itemPipe): {
            this->backup();
            return;
        }
        case (itemSemicolon): {
            return;
        }
        case (itemNL): {
            LexTokenItemTyp peek_type = this->peek().typ;
            if (peek_type == itemPipe || peek_type == itemNL || peek_type == itemSemicolon) {
                return;
            }
            break;
        }
        case (itemIdent): {
            e->tokens.push_back(get_new_token_node(cur_token.val, TypTermKeyword));
            break;
        }
        case (itemLiteral): {
            e->tokens.push_back(get_new_token_node(cur_token.val, TypLiteral));
            break;
        }
        case (itemExpr): {
            e->command = cur_token.val;
            if (this->peek().typ == itemNL) {
                this->next();
            }
            break;
        }
        case (itemPct):
            [[fallthrough]];
        case (itemComment):
            break;
        default: {
            this->unexpected(cur_token, context);
        }
        }
    }
}

vector<ExpressionNode*> Tree::parse_expression_lemon(int pos)
{
    const string context = "expression";
    vector<ExpressionNode*> v_e;
    v_e.push_back(get_new_expression(pos));

    for (;;) {
        const LexTokenItem cur_token = this->next();
        switch (cur_token.typ) {
        case (itemNL): {
            // all expressions are in one line
            // as formatted by our generate_ir_types_sqlite.py.
            // Return empty new expression.
            return v_e;
        }
        case (itemIdent): {
            for (auto* e : v_e) {
                e->tokens.push_back(get_new_token_node(cur_token.val, TypTermKeyword));
            }
            break;
        }
        case (itemLiteral): {
            for (auto* e : v_e) {
                e->tokens.push_back(get_new_token_node(cur_token.val, TypLiteral));
            }

            break;
        }
        case (itemExpr): {
            for (auto* e : v_e) {
                e->command = cur_token.val;
            }
            if (this->peek().typ == itemNL) {
                this->next();
            }
            return v_e;
        }
        case (itemComment):
            break;
        case (itemTerm): {
            // if encounter te termination period, ignore all other text until the end of line.
            for (LexTokenItem next_token = this->next(); next_token.typ != itemNL && next_token.typ != itemEOF; next_token = this->next()) { }
            // Backup the last new line of EOF token.
            this->backup();
            return v_e;
        }
        case (itemPipe): {
            ExpressionNode* new_expr_node = v_e.back()->deep_copy();

            // If encounter itemPipe, meaning that there is one token that have already
            // been duplicate pushed to both previous and this new expression. Remove it.
            TokenNode* node_to_rov = new_expr_node->tokens.back();
            new_expr_node->tokens.pop_back();
            delete node_to_rov;

            const LexTokenItem pipe_next_token = this->next();

            if (pipe_next_token.typ == itemIdent) {
                new_expr_node->tokens.push_back(get_new_token_node(pipe_next_token.val, TypTermKeyword));
            } else if (pipe_next_token.typ == itemLiteral) {
                new_expr_node->tokens.push_back(get_new_token_node(pipe_next_token.val, TypLiteral));
            }

            v_e.push_back(new_expr_node);

            break;
        }
        default: {
            this->unexpected(cur_token, context);
        }
        }
    }
}